dev-tools/systemtests/container.go

```diff
@@ -0,0 +1,303 @@
+	defer reader.Close()
+
+	_, err = io.Copy(os.Stdout, reader)
+	require.NoError(tr.Runner, err, "error copying image")
+
+	wdCmd := exec.Command("git", "rev-parse", "--show-toplevel")
+	wdPath, err := wdCmd.CombinedOutput()
+	require.NoError(tr.Runner, err, "error finding root path")
+
+	cwd := strings.TrimSpace(string(wdPath))
+	logp.L().Infof("using cwd: %s", cwd)
+
+	testRunCmd := []string{"go", "test", "-v", tr.Basepath}
+	if tr.Testname != "" {
+		testRunCmd = append(testRunCmd, "-run", tr.Testname)
+	}
+
+	mountPath := "/hostfs"
+
+	containerEnv := []string{fmt.Sprintf("HOSTFS=%s", mountPath)}
+	if tr.Privileged {
+		containerEnv = append(containerEnv, "PRIVILEGED=1")
+	}
+
+	if tr.MonitorPID != 0 {
+		containerEnv = append(containerEnv, fmt.Sprintf("MONITOR_PID=%d", tr.MonitorPID))
+	}
+
+	resp, err := apiClient.ContainerCreate(ctx, &container.Config{
+		Image:      tr.Container,
+		Cmd:        testRunCmd,
+		Tty:        false,
+		WorkingDir: "/app",
+		Env:        containerEnv,
+		User:       tr.RunAsUser,
+	}, &container.HostConfig{
+		CgroupnsMode: tr.CgroupNSMode,
+		Privileged:   tr.Privileged,
+		Binds:        []string{fmt.Sprintf("/:%s", mountPath), fmt.Sprintf("%s:/app", cwd)},
+	}, nil, nil, "")
+	require.NoError(tr.Runner, err, "error creating container")
+
+	return resp
+}
+
+func (tr *DockerTestRunner) runContainerTest(ctx context.Context, apiClient *client.Client, resp container.CreateResponse) RunResult {
+	err := apiClient.ContainerStart(ctx, resp.ID, container.StartOptions{})
+	require.NoError(tr.Runner, err, "error starting container")
+
+	res := RunResult{}
+
+	statusCh, errCh := apiClient.ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
+	select {
+	case err := <-errCh:
+		require.NoError(tr.Runner, err, "error in container")
+	case status := <-statusCh:
+		res.ReturnCode = status.StatusCode
+	}
+
+	out, err := apiClient.ContainerLogs(ctx, resp.ID, container.LogsOptions{ShowStdout: true, ShowStderr: true})
+	require.NoError(tr.Runner, err, "error fetching logs")
+
+	stdout := bytes.NewBufferString("")
+	stderr := bytes.NewBufferString("")
+	_, err = stdcopy.StdCopy(stdout, stderr, out)
+	require.NoError(tr.Runner, err, "error copying logs")
+	res.Stderr = stderr.String()
+	res.Stdout = stdout.String()
+
+	return res
+}
+
+func (tr *DockerTestRunner) createMonitoredProcess(ctx context.Context) {
+	log := logp.L()
+	// if user has specified a process to monitor, start it now
+	// skip if the process has already been created
+	if tr.CreateHostProcess != nil && tr.CreateHostProcess.Process == nil {
+		// We don't need to do this in a channel, but it prevents races between this goroutine
+		// and the rest of test framework
+		startPid := make(chan int)
+		log.Infof("Creating test Process...")
+		go func() {
+			err := tr.CreateHostProcess.Start()
+			// if the process fails to start up, the resulting tests will fail, so just log it
+			require.NoError(tr.Runner, err, "error starting monitor process")
```

Comment:
`require` uses `t.FailNow` which should not be called from a different goroutine than the one running the test, from the [docs](https://pkg.go.dev/testing#T.FailNow):
> (...) FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. Calling FailNow does not stop those other goroutines.

It's better to use `assert` here
```suggestion
			assert.NoError(tr.Runner, err, "error starting monitor process")
```